Content.Server/Replays/ReplayEventSystem.cs

```diff
@@ -0,0 +1,245 @@
+    }
+
+    private void OnRecordEventsChanged(bool enabled)
+    {
+        _recordEvents = enabled;
+    }
+
+    private void ReplaysOnRecordingStarted(MappingDataNode arg1, List<object> arg2)
+    {
+        _replayEvents = new List<ReplayEvent>();
+    }
+
+    private void ReplaysOnRecordingStopped(ReplayRecordingStopped replayRecordingStopped)
+    {
+        var events = _serialman.WriteValue(_replayEvents, true, null);
+        var bytes = Encoding.UTF8.GetBytes(events.ToString());
+        replayRecordingStopped.Writer.WriteBytes(replayRecordingStopped.Writer.BaseReplayPath / "events.yml", new ReadOnlyMemory<byte>(bytes));
+    }
+
+    /// <summary>
+    /// Records a replay event. This is the main way to record events in the replay system.
+    /// </summary>
+    /// <param name="replayEvent">The event to record</param>
+    /// <param name="source">Optional source that will be used for location data</param>
+    public void RecordReplayEvent(ReplayEvent replayEvent, EntityUid? source = null)
+    {
+        if (!_replays.IsRecording)
+            return;
+
+        replayEvent.Time ??= _gameTiming.CurTime.TotalSeconds;
+
+        if (source.HasValue)
+        {
+            replayEvent.Position = _transformSystem.GetWorldPosition(source.Value);
+            replayEvent.NearestBeacon =
+                _navMapSystem.GetNearestBeaconString(_transformSystem.GetMapCoordinates(source.Value));
+
+            var map = _transformSystem.GetMap(source.Value);
+            if (map.HasValue)
+            {
+                replayEvent.Map = EntityManager.GetComponent<MetaDataComponent>(map.Value).EntityName;
+            }
+        }
+
+        DebugTools.AssertNotNull(replayEvent.EventType);
+        DebugTools.AssertNotNull(replayEvent.Severity);
+
+        Log.Debug($"Recording replay event: {replayEvent.EventType}");
+        if (_replayEvents == null)
+        {
+            // If this happens, someone messed up.
+            Log.Error("Tried to record a replay event, but the events list is null. This should never happen.");
+            return;
+        }
+        _replayEvents.Add(replayEvent);
+    }
+
+    /// <summary>
+    /// Gets the player info for a player entity.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">
+    /// Thrown if the entity is not a player.
+    /// </exception>
+    public ReplayEventPlayer GetPlayerInfo(EntityUid player)
+    {
+        if (EntityManager.TryGetComponent<ActorComponent>(player, out var actorComponent))
+        {
+            return GetPlayerInfo(actorComponent.PlayerSession);
+        }
+
+        Log.Warning($"Tried to get player info for entity {player}, but it's not a player entity.");
+        return new ReplayEventPlayer()
+        {
+            PlayerGuid = new NetUserId(Guid.Empty),
+            PlayerICName = EntityManager.GetComponent<MetaDataComponent>(player).EntityName, // Fallback, best we can do.
+            PlayerOOCName = "Unknown",
```

Comment:
"Unknown" is an actual valid account name, so having it like this leaves a real possibility of somebody accidentally misattributing the event. Id rather just leave this blank.